### PR TITLE
Multiple fixes over the reservation patch.

### PR DESCRIPTION
**Pull request by qwattash**
This addresses most post-merge comments from @bsdjhb for #614

### CODE REVIEW

**Review by bsdjhb — APPROVED** [review]

File: sys/dev/pci/pci_user.c [hunk 1]
```diff
@@ -1403,6 +1427,18 @@ pci_ioctl(struct cdev *dev, u_long cmd, caddr_t data, int flag, struct thread *t
...
+			pbm64 = (struct pci_bar_mmap64 *)data;
+			pbm = &pbms;
+			pbm->pbm_map_base = __USER_CAP(pbm64->pbm_map_base,
+			    pbm64->pbm_map_length);
+			memcpy(__unbounded_addressof(pbm->pbm_map_length),
+			    __unbounded_addressof(pbm64->pbm_map_length),
+			    sizeof(struct pci_bar_mmap64) - sizeof(uint64_t));
+		}
+#endif
 		pbm = (struct pci_bar_mmap *)data;
```
Comment (brooksdavis) [inline sys/dev/pci/pci_user.c:1442]:
```suggestion
		} else
#endif
			pbm = (struct pci_bar_mmap *)data;
```

File: sys/dev/pci/pci_user.c [hunk 2]
```diff
@@ -1403,6 +1427,18 @@ pci_ioctl(struct cdev *dev, u_long cmd, caddr_t data, int flag, struct thread *t
...
+#ifdef COMPAT_FREEBSD64
+	case PCIOCBARMMAP64:
+		if (cmd == PCIOCBARMMAP64) {
+			pbm64 = (struct pci_bar_mmap64 *)data;
+			pbm = &pbms;
+			pbm->pbm_map_base = __USER_CAP(pbm64->pbm_map_base,
+			    pbm64->pbm_map_length);
+			memcpy(__unbounded_addressof(pbm->pbm_map_length),
+			    __unbounded_addressof(pbm64->pbm_map_length),
+			    sizeof(struct pci_bar_mmap64) - sizeof(uint64_t));
```
Comment (brooksdavis) [inline sys/dev/pci/pci_user.c:1439]:
```suggestion
			    sizeof(struct pci_bar_mmap64) - offsetof(struct pci_bar_mmap64, pbm_map_length);
```

Comment (brooksdavis) [reply sys/dev/pci/pci_user.c:1439]: That being said, the usual style would be to copy individual fields with
```C
CP(pbm64, pbm, pbm_map_length);
CP(pbm64, pbm, pbm_sel);
...
```

**Review by brooksdavis — CHANGES_REQUESTED** [review]
pci_bar_mmap modifies the `pbm` argument and it's copied back out to userspace so you need to update `data` after a successful call.

File: sys/dev/pci/pci_user.c (hunk 2 above)
Comment (qwattash) [reply sys/dev/pci/pci_user.c:1439]: Moved to `CP()` style copy.

File: sys/dev/pci/pci_user.c [hunk 3]
```diff
@@ -986,7 +1000,7 @@ pci_bar_mmap(device_t pcidev, struct pci_bar_mmap *pbm)
 	flags = MAP_SHARED;
 	addr = 0;
 	if ((pbm->pbm_flags & PCIIO_BAR_MMAP_FIXED) != 0) {
-		addr = (uintptr_t)pbm->pbm_map_base;
+		addr = (__cheri_addr vm_ptr_t)pbm->pbm_map_base;
 		flags |= MAP_FIXED;
```
Comment (brooksdavis) [inline sys/dev/pci/pci_user.c:1004]: Hmm, this code needs the whole mmap addr treatment (MAP_FIXED requires a VMMAP bearing cap or a NULL-derived cap which implies MAP_EXCLUSIVE).

Comment (qwattash) [reply sys/dev/pci/pci_user.c:1023]: Should be fixed now

File: sys/dev/pci/pci_user.c [hunk 4]
```diff
@@ -972,6 +987,25 @@ pci_bar_mmap(device_t pcidev, struct pci_bar_mmap *pbm)
...
+	flags = MAP_SHARED;
+#if __has_feature(capabilities)
+	/* Enforce the same policy as for sys_mmap() */
+	if (cheri_gettag(pbm->pbm_map_base)) {
+		if ((pbm->pbm_flags & PCIIO_BAR_MMAP_FIXED) == 0)
+			return (EPROT);
+		else if ((cheri_getperm(pbm->pbm_map_base) &
+		    CHERI_PERM_CHERIABI_VMMAP) == 0)
+			return (EACCES);
+	}
```
Comment (bsdjhb) [inline sys/dev/pci/pci_user.c:999]: Minor style nit, place the `else {` on the same line as the closing `}`.

File: sys/dev/pci/pci_user.c [hunk 5]
```diff
@@ -987,6 +987,24 @@ pci_bar_mmap(device_t pcidev, struct pci_bar_mmap *pbm)
 	prot = VM_PROT_READ | (((pbm->pbm_flags & PCIIO_BAR_MMAP_RW) != 0) ?
 	    VM_PROT_WRITE : 0);
 
+	flags = MAP_SHARED;
+#if __has_feature(capabilities)
+	/* Enforce the same policy as for sys_mmap() */
+	if (cheri_gettag(pbm->pbm_map_base)) {
+		if ((pbm->pbm_flags & PCIIO_BAR_MMAP_FIXED) == 0)
+			return (EPROT);
+		else if ((cheri_getperm(pbm->pbm_map_base) &
```
Comment (brooksdavis) [inline sys/dev/pci/pci_user.c:996]: no need for `else` when the body above is a `return`

**Comment by qwattash** [comment]
@brooksdavis If you are happy I can merge now.

**Review by brooksdavis — APPROVED** [review]

**Pull request merged**